### ISSUE TITLE
include/extends in pug

### ISSUE THREAD

**Issue by Ferym26** [pugjs/pug#2292]
Why 'includes' and 'extend' dont work in pug like it worked in jade?
I cant configure paths in them.

my progect:
workstorm.local
=app
=frontend
==pug
==index.pug
===includes
====layout.pug
====header.pug

**"gulp-pug": "^1.0.0"**
**"pug": "^2.0.0-alpha1"**

gulp task
``` js
gulp.task('pug', function build() {
  return gulp.src('frontend/pug/**/*.pug')
  	.pipe(debug({title: 'src'}))
    .pipe(pug({pretty: true, filename: 'workstorm.local/'}))
    .pipe(debug({title: 'pug'}))
    .pipe(gulp.dest('app'));
});
```

index.pug
```
extend frontend/pug/includes/layout.pug

block content
	// USER CONTENT
	h1 WWWHHYYYYYYYYYYYYYY????
	p bobobob
	// END USER CONTENT
```

layout.pug
```
...
	body
		
		include /includes/header.pug

		block content
...
```

header.pug
```
// HEADER
p header
//End HEADER
```

And every time I get an error like this: 
[error scrin](http://c2n.me/3wwIZCq)
Error: ENOENT: no such file or directory, open 'f:\PROJECTS\HTML\Pagelayout\workstorm.local\frontend\header.pug'
with several paths
I tested all the options of paths in "filename: 'workstorm.local/'" and "include ..." "extend..."
what am I doing wrong? thx

**Comment by ForbesLindesay** [comment]
Does 'f:\PROJECTS\HTML\Pagelayout\workstorm.local\frontend\header.pug' exist?  Is it resolving to the wrong path, or failing to load the path once correctly resolved?

It looks like you're using absolute filepaths (i.e. starting with `/`) but you haven't specified a `basedir` option.  This should be throwing an error like 'the "basedir" option is required to use includes and extends with "absolute" paths'.  Can you perhaps provide an example as a GitHub repo or zip file somewhere so I can attempt to reproduce it?

**Comment by Ferym26** [comment]
Here it is https://github.com/Ferym26/workstorm

**Comment by pateketrueke** [comment]
I have a bunch of sources using `.jade`extensions but `include example.jade` doesn't work for me, it just output the pug source; when I rename it to use `include example.pug` it works, I don't have such control over third-party sources, there's no legacy support for this?

**Comment by ForbesLindesay** [comment]
@pateketrueke Unfortunately, you'll need to rename them all to `.pug` currently, I guess we could consider some kind of legacy support for `.jade` files, perhaps with a warning or something.

@Ferym26 You're running into jamen/gulp-pug#4 is going to be a bit of an issue for you.  The `filename` option should always be the full file path of the template being compiled, but you're setting it to a constant value.  Relative paths can then be relative to the current file.  You can alternatively use absolute paths (N.B. leading `/` on paths):

in index.pug

```diff
- extend frontend/pug/includes/layout.pug
+ extend /includes/layout.pug
```

Then in your gulpfile:

```diff
-    .pipe(pug({pretty: true, filename: 'workstorm.local/'}))
+    .pipe(pug({pretty: true, basedir: __dirname + '/frontend/pug'}))
```

Alternatively, if jamen/gulp-pug#4 gets fixed, you could use relative paths (N.B. no leading `/` on paths):


```diff
- extend frontend/pug/includes/layout.pug
+ extend includes/layout.pug
```

In your layout.pug file

```diff
- include /includes/header.pug
+ include header.pug
```

Then in your gulpfile:

```diff
-    .pipe(pug({pretty: true, filename: 'workstorm.local/'}))
+    .pipe(pug({pretty: true}))
```

This issue isn't actually new to pug and would apply equally if you were using the latest version of jade, so it's not a regression.  It may be a regression in `gulp-pug` vs. `gulp-jade`, but I don't maintain those.

**Comment by jamen** [comment]
@Ferym26: I've fixed the importing for gulp-pug in v1.1 (jamen/gulp-pug#6).  `filename` is set per file, so the including/extending is relative to the file.

**Comment by ORESoftware** [comment]
I am getting:

> Error: the "filename" option is required to use includes and extends with "relative" paths
>     at undefined line 1


and I am using:

```javascript
{
  const templateStr = readViewPath('index.pug');
  const fn = pug.compile(templateStr, {baseDir: pugBaseDir});
  const html = fn({title: 'Main'});
  router.get('*', setStandardHeaders(html.length), (req, res) => {
    res.send(html);
  });
}

```

shouldn't basedir bet enough?
